wasm/src/main/scala/ir2wasm/WasmExpressionBuilder.scala

```diff
@@ -2515,4 +2449,430 @@ private class WasmExpressionBuilder private (
+           *   end
+           *   ref.null exn
+           * end
+           */
+          if (entry.wasCrossed) {
+            val (destinationTagLocal, crossLabel) = entry.requireCrossInfo()
+
+            // Go back and insert the `block $cross` right after `block $catch`
+            instrs.insert(
+              instrsBlockBeginIndex,
+              BLOCK(BlockType.ValueType(), Some(crossLabel))
+            )
+
+            // And the other amendments normally
+            instrs += I32_CONST(0)
+            instrs += LOCAL_SET(destinationTagLocal)
+            instrs += END // of the inserted BLOCK
+          }
+
+          // on success, push a `null_ref exn` on the stack
+          instrs += REF_NULL(Types.WasmHeapType.Exn)
+        } // end block $catch
+
+        // finally block (during which we leave the `(ref null exn)` on the stack)
+        genTree(t.finalizer, IRTypes.NoType)
+
+        if (!entry.wasCrossed) {
+          // If the `exnref` is non-null, rethrow it
+          instrs += BR_ON_NULL(doneLabel)
+          instrs += THROW_REF
+        } else {
+          // If the `exnref` is non-null, rethrow it, but stay within the `$done` block
+          fctx.block(WasmFunctionSignature(List(Types.WasmRefType.exnref), Nil)) {
+            exnrefIsNullLabel =>
+              instrs += BR_ON_NULL(exnrefIsNullLabel)
+              instrs += THROW_REF
+          }
+
+          /* Otherwise, use a br_table to dispatch to the right destination
+           * based on the value of the function's destinationTagLocal.
+           */
+
+          // The order does not matter here because they will be "re-sorted" by emitBRTable
+          val possibleTargetEntries =
+            enclosingLabeledBlocks.valuesIterator.filter(_.wasCrossUsed).toList
+
+          val nextTryFinallyEntry = innermostTryFinally // note that we're out of ourselves already
+            .filter(nextTry => possibleTargetEntries.exists(nextTry.isInside(_)))
+
+          /* Build the destination table for `br_table`. Target Labeled's that
+           * are outside of the next try..finally in line go to the latter;
+           * for other `Labeled`'s, we go to their cross label.
+           */
+          val brTableDests: List[(Int, WasmLabelName)] = possibleTargetEntries.map { targetEntry =>
+            val (destinationTag, _, crossLabel) = targetEntry.requireCrossInfo()
+            val label = nextTryFinallyEntry.filter(_.isInside(targetEntry)) match {
+              case None          => crossLabel
+              case Some(nextTry) => nextTry.requireCrossInfo()._2
+            }
+            destinationTag -> label
+          }
+
+          instrs += LOCAL_GET(entry.requireCrossInfo()._1)
+          for (nextTry <- nextTryFinallyEntry) {
+            // Transfer the destinationTag to the next try..finally in line
+            instrs += LOCAL_TEE(nextTry.requireCrossInfo()._1)
+          }
+          emitBRTable(brTableDests, doneLabel)
```

Comment:
[solved] 😅 

Added test that the fall-through is executed before the outer finally block

```diff
diff --git a/test-suite/src/main/scala/testsuite/core/TryFinallyReturnTest.scala b/test-suite/src/main/scala/testsuite/core/TryFinallyReturnTest.scala
index 9a71700..2b3d15c 100644
--- a/test-suite/src/main/scala/testsuite/core/TryFinallyReturnTest.scala
+++ b/test-suite/src/main/scala/testsuite/core/TryFinallyReturnTest.scala
@@ -45,6 +45,12 @@ Running nestedFinallyBlocks
 in finally 1
 in finally 2
 ----------------------------------------
+Running nestedFinallyBlocks2
+in try 1
+in finally 1
+in fall-through
+in finally 2
+----------------------------------------
 """
 
   def main(): Unit = {
@@ -56,6 +62,7 @@ in finally 2
     test(retFinally(), "retFinally")
     test(throwFinally(), "throwFinally")
     test(nestedFinallyBlocks(), "nestedFinallyBlocks")
+    test(nestedFinallyBlocks2(), "nestedFinallyBlocks2")
 
     assertSame(expectedOutput, printlnOutput)
   }
@@ -162,6 +169,19 @@ in finally 2
       println("in finally 2")
     }
 
+  def nestedFinallyBlocks2(): Int =
+    try {
+      try {
+        println("in try 1")
+      } finally {
+        println("in finally 1")
+      }
+      println("in fall-through")
+      0
+    } finally {
+      println("in finally 2")
+    }
+
   def test[A](m: => A, name: String): Unit = {
     println("Running %s".format(name))
     try {
```

and this test passed. While I assumed that this test will fail because it seems that we anyway overwrite the destinationTag to the outer finally block instead of the fall-through.

How this test is passing?

Reply:
Oops, never mind, `possibleTargetEntries` doesn't contains the fall-thgrough

Reply:
I added the test anyway. ;)